gremlin-dotnet/src/Gremlin.Net/Driver/AsyncAutoResetEvent.cs

```diff
@@ -0,0 +1,103 @@
+﻿#region License
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+// The implementation is based on this blog post by Stephen Toub:
+// https://blogs.msdn.microsoft.com/pfxteam/2012/02/11/building-async-coordination-primitives-part-2-asyncautoresetevent/
+
+namespace Gremlin.Net.Driver
+{
+    /// <summary>
+    ///     An async version of the AutoResetEvent.
+    /// </summary>
+    public class AsyncAutoResetEvent
+    {
+        private static readonly Task<bool> CompletedTask = Task.FromResult(true);
+        private readonly List<TaskCompletionSource<bool>> _waitingTasks = new List<TaskCompletionSource<bool>>();
+        private bool _isSignaled;
+        
+        /// <summary>
+        ///     Asynchronously waits for this event to be set or until a timeout occurs.
+        /// </summary>
+        /// <param name="timeout">A <see cref="TimeSpan"/> that that represents the number of milliseconds to wait.</param>
+        /// <returns>true if the current instance received a signal before timing out; otherwise, false.</returns>
+        public async Task<bool> WaitOneAsync(TimeSpan timeout)
+        {
+            var tcs = new TaskCompletionSource<bool>();
+            var waitTask = WaitForSignalAsync(tcs);
+            if (waitTask.IsCompleted) return true;
+            
+            await Task.WhenAny(waitTask, Task.Delay(timeout)).ConfigureAwait(false);
```

Comment:
Each `Task.Delay()` uses an individual timer that are [a limited resource](https://blogs.msdn.microsoft.com/oldnewthing/20090827-00/?p=16963/).

A workaround would be to have a single timer that ticks at certain interval (e.g: 100ms) and checks the queue of tasks.

Reply:
Are you sure that that is still true? I just did a quick test and executed 1 Mio. times `Task.Delay` in parallel with a 10 second delay and the whole test finished in 11 seconds. Unfortunately I couldn't find any more up-to-date information about this.

The full test for reference:
```cs
[Fact]
public async Task TimersTest()
{
    const int nrTasks = 1000000;
    var tasks = new List<Task>(nrTasks);
    for (var i = 0; i < nrTasks; i++)
    {
        tasks.Add(Task.Delay(TimeSpan.FromSeconds(10)));
    }

    await Task.WhenAll(tasks);
}
```

Reply:
hmm... indeed, it looks like in .NET Core it uses a single native timer for all timer instances: https://github.com/dotnet/coreclr/blob/master/src/System.Private.CoreLib/src/System/Threading/Timer.cs

I wouldn't know about other runtimes.

Reply:
I just tried the same test with .NET Framework 4.6 and got the same result. This is also in line with [the reference source for .NET Framework 4.7.2](https://referencesource.microsoft.com/#mscorlib/system/threading/timer.cs). Mono seems to use this reference source as a basis [for their implementation](https://github.com/mono/mono/blob/c5b88ec4f323f2bdb7c7d0a595ece28dae66579c/mcs/class/referencesource/mscorlib/system/threading/timer.cs).

Using `Task.Delay()` together with `Task.WhenAny()` to time out tasks is also a pattern recommended in the two resources you linked to earlier. So, it would be really strange if it worked only for a small number of tasks.

Are you still in favour of a more complicated implementation like such a timer queue you mentioned?